dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -63,11 +64,14 @@ public abstract class KernelFunction
     /// <param name="name">Name of the function.</param>
     /// <param name="description">Function description.</param>
     /// <param name="modelSettings">Model request settings.</param>
-    internal KernelFunction(string name, string description, IEnumerable<AIRequestSettings>? modelSettings = null)
+    /// <param name="invokeEventHandlers">Flad to control whether or not event handlers are invoked</param>
+    internal KernelFunction(string name, string description, IEnumerable<AIRequestSettings>? modelSettings = null, bool invokeEventHandlers = true)
```

Comment:
Why `inovkeEventHandlers` flag?

IMO all functions should invoke...

Reply:
There is because Plan wraps steps (which are KernelFunctions) in Plans (which are KernelFunctions) and I have to avoid triggering the hooks twice. The proper fix would be to change Plan to had a collection of Plans | KernelFunctions but that's too involved and the future of Plan is uncertain.

Reply:
> and I have to avoid triggering the hooks twice

Why? Wouldn't we want any function invocation to trigger the events whether it wraps other functions or not?

Reply:
In this case I don't think we should be wrapping the individual steps. I'll undo the change and update the tests the account for the extra events.